src/test/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorServiceTest.java

```diff
@@ -24,22 +24,97 @@
 import org.junit.Rule;
+import org.junit.rules.ErrorCollector;
 import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.Issue;
 import org.jvnet.hudson.test.JenkinsRule;
 
 public class CpsVmExecutorServiceTest {
 
     @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
     @Rule public JenkinsRule r = new JenkinsRule();
+    @Rule public ErrorCollector errors = new ErrorCollector();
 
     @Test public void contextClassLoader() throws Exception {
         WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
         p.setDefinition(new CpsFlowDefinition("echo(/yes I can load ${Thread.currentThread().contextClassLoader.loadClass(getClass().name)}/)", false));
         r.buildAndAssertSuccess(p);
     }
 
+    @Issue({"JENKINS-31314", "JENKINS-27306"})
+    @Test public void wrongCatcher() throws Exception {
+        boolean origFailOnMismatch = CpsVmExecutorService.FAIL_ON_MISMATCH;
+        CpsVmExecutorService.FAIL_ON_MISMATCH = false;
+        try {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("def ok() {sleep 1}; @NonCPS def bad() {for (int i = 0; i < 10; i++) {sleep 1}; assert false : 'never gets here'}; node {ok(); bad()}", true));
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("WorkflowScript", "bad", null, "sleep"), r.buildAndAssertSuccess(p));
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("def l = [3, 2, 1]; println(/oops got ${l.sort {x, y -> x - y}}/)", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains("oops got -1", b);
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("java.util.ArrayList", "sort", "org.jenkinsci.plugins.workflow.cps.CpsClosure2", "call"), b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("node {[1, 2, 3].each {x -> sleep 1; echo(/no problem got $x/)}}", true));
+                WorkflowRun b = r.buildAndAssertSuccess(p);
+                r.assertLogContains("no problem got 3", b);
+                r.assertLogNotContains("expected to call", b);
+                return null;
+            });
+            errors.checkSucceeds(() -> {
+                p.setDefinition(new CpsFlowDefinition("class C {@Override String toString() {'never used'}}; def gstring = /embedding ${new C()}/; echo(/oops got $gstring/)", true));
+                WorkflowRun b = r.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0)); // JENKINS-27306: No such constructor found: new org.codehaus.groovy.runtime.GStringImpl java.lang.String java.lang.String[]
+                r.assertLogContains(CpsVmExecutorService.mismatchMessage("java.lang.Class", "asType", "C", "toString"), b);
```

Comment:
The expected receiver class name seems a bit weird here. If you patch `CpsVmExecutorService.className` to return the class name directly when the expected receiver is a class and not a method rather than calling `getClass` on the class then you get the following, which seems better:

> expected to call org.codehaus.groovy.runtime.ScriptBytecodeAdapter.asType but wound up catching C.toString

```
diff --git a/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java b/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java
index d094278e..7b7691d5 100644
--- a/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java
+++ b/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java
@@ -125,6 +125,9 @@ class CpsVmExecutorService extends InterceptingExecutorService {
     }
 
     private static @CheckForNull String className(@CheckForNull Object receiver) {
+        if (receiver instanceof Class) {
+            return ((Class) receiver).getName();
+        }
         return receiver == null ? null : receiver.getClass().getName();
     }
```